### PR TITLE
Enforce parity between Jsonnet and Helm

### PR DESCRIPTION
**Pull request by Logiraptor**
#### What this PR does

Use kustomize to compare Jsonnet with Helm automatically.

See the section in [docs](https://github.com/grafana/mimir/blob/logiraptor/use-k8s-diff/docs/internal/contributing/contributing-to-helm-chart.md#automated-comparison-with-jsonnet) for more information.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [x] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
+package main
```
Comment (dimitarvdimitrov) [inline operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:1]: would be nice to document what this program does. A short readme should suffice

**Review by dimitarvdimitrov — COMMENTED** [review]
that's how far i got today

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go (hunk 1 above)
Comment (Logiraptor) [reply operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:1]: Agreed, I'll take care of it

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/full-input.yaml [hunk 2]
```diff
@@ -0,0 +1,2781 @@
+apiVersion: config.kubernetes.io/v1
```
[the remark this answers is not in the document]
Comment (Logiraptor) [reply operations/compare-helm-with-jsonnet/plugins/resolve-config/full-input.yaml:1]: On second thought, I'll replace it with some actual tests.

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+// copied here in order to avoid a dependency on the k8s.io/api/apps/v1 package
+type DeploymentOrStatefulSet struct {
+	Spec struct {
+		Template struct {
+			Spec PodSpec `yaml:"spec"`
+		} `yaml:"template"`
+	} `yaml:"spec"`
+}
+
+func (c *ConfigExtractor) ExtractPodSpec(obj *yaml.RNode) (PodSpec, bool, error) {
```
Comment (dimitarvdimitrov) [inline operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:96]: Any reason for keeping this a method vs a regular function?

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+	var mimirConfig = &mimir.Config{}
+
+	flagSet := flag.NewFlagSet("mimir", flag.ContinueOnError)
+	// Suppress usage output
+	flagSet.Usage = func() {}
+
+	// This sets default values on the config struct, so it needs to be called before parsing the config file
+	mimirConfig.RegisterFlags(flagSet, log.NewNopLogger())
+	flagSet.Bool("config.expand-env", false, "")
+	flagSet.Int("mem-ballast-size-bytes", 0, "")
```
Comment (dimitarvdimitrov) [inline operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:203]: Why are these two needed?

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+		rawConfigs = append(rawConfigs, r)
+	}
+
+	return rawConfigs, nil
+}
+
+func (c *ConfigExtractor) resolveArgsAndConfigFile(args []string, configFileText string) (*yaml.RNode, string, error) {
+	var mimirConfig = &mimir.Config{}
+
+	flagSet := flag.NewFlagSet("mimir", flag.ContinueOnError)
```
Comment (dimitarvdimitrov) [inline operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:196]: I'm not sure about `ContinueOnError`. Shouldn't all flags successfully marshal into a mimir config?

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/main_test.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+		target, err := configs[0].GetFieldValue("config.target")
+		require.NoError(t, err)
+		require.Equal(t, "querier", target, "The MimirConfig should have its target set to the configured target")
+
+		bindPort, err := configs[0].GetFieldValue("config.memberlist.bind_port")
+		require.NoError(t, err)
+		require.Equal(t, "7946 (default)", bindPort, "The MimirConfig should have its memberlist bind port set to the default value, with the text 'default'")
+	})
+
+	t.Run("ResolveConfigs sets configuration parameters from cli arguments", func(t *testing.T) {
```
Comment (dimitarvdimitrov) [inline operations/compare-helm-with-jsonnet/plugins/resolve-config/main_test.go:48]: nit: this + the following 2 tests can be converted into a table test, they have the same assertions.

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, it's a hefty piece of work

File: docs/internal/contributing/contributing-to-helm-chart.md [hunk 7]
```diff
@@ -33,3 +33,81 @@ Install [ct](https://github.com/helm/chart-testing) and run
 ```bash
 ct lint --config operations/helm/ct.yaml --charts operations/helm/charts/mimir-distributed
 ```
+
+## Automated Comparison with Jsonnet
```
Comment (osg-grafana) [inline docs/internal/contributing/contributing-to-helm-chart.md:37]:
```suggestion
## Automated comparison with Jsonnet
```

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go (hunk 3 above)
Comment (Logiraptor) [reply operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go]: nope, updated

File: operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go [hunk 8]
```diff
@@ -0,0 +1,349 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
```
[the remark this answers is not in the document]
Comment (Logiraptor) [reply operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:5]: I ran gofmt and goimports -local, is there some other format I'm not aware of?

[the remark this answers is not in the document]
Comment (56quarters) [reply operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:5]: Normally (could have sworn this was enforced by lint), imports in Mimir are grouped:
* stdlib
* 3rd party
* this project

[the remark this answers is not in the document]
Comment (56quarters) [reply operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:5]: So, `"github.com/pkg/errors"` would move to the 3rd party group

[the remark this answers is not in the document]
Comment (Logiraptor) [reply operations/compare-helm-with-jsonnet/plugins/resolve-config/main.go:5]: Thanks! I removed the newline and my editor rearranged it properly.

**Comment by krajorama** [comment]
trivial rebase

File: operations/compare-helm-with-jsonnet/helm/06-pods/kustomization.yaml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+  # TODO(logiraptor): Helm adds an emptyDir volume to several components. This is likely due to the fact that readOnlyRootFilesystem=true and some ephemeral data is written
+  - target:
+      name: mimir-(distributor|query-frontend|querier|ruler|querier|overrides-exporter|alertmanager)
+      kind: Deployment|StatefulSet
+    patch: |-
+      - op: remove
+        path: /spec/template/spec/volumes/1
+
+  # Helm attempts to automatically discover which PDB apiVersion is supported,
```
Comment (krajorama) [inline operations/compare-helm-with-jsonnet/helm/06-pods/kustomization.yaml:42]: FYI, In the `ct` tests we circumvent this "flakyness" by setting the kubernetes version explicitly: 

https://github.com/grafana/mimir/blob/23a9fe46b53fc30ab8e3fdecf224c7033d04c2fb/operations/helm/charts/mimir-distributed/ci/test-oss-values.yaml#L2

File: Makefile [hunk 10]
```diff
@@ -509,7 +509,13 @@ check-jsonnet-getting-started:
 		| sed 's/\(jb install github.com\/grafana\/mimir\/operations\/mimir@main\)/\1 \&\& rm -fr .\/vendor\/mimir \&\& cp -r ..\/operations\/mimir .\/vendor\/mimir\//g' \
 		| bash
 
-build-helm-tests:
+operations/helm/charts/mimir-distributed/charts: operations/helm/charts/mimir-distributed/Chart.yaml operations/helm/charts/mimir-distributed/Chart.lock
+	@cd ./operations/helm/charts/mimir-distributed && helm dependency update
+
+check-helm-jsonnet-diff: operations/helm/charts/mimir-distributed/charts
```
Comment (krajorama) [inline Makefile:515]: Shouldn't we depend on the jsonnet base to be built?

```suggestion
check-helm-jsonnet-diff: operations/helm/charts/mimir-distributed/charts build-jsonnet-tests
```

**Review by krajorama — COMMENTED** [review]
Some nits and a question about the Makefile

**Pull request merged**